src/coreclr/jit/importer.cpp

```diff
@@ -3021,6 +3021,25 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
                                                typeInfo(TYP_INT));
                                 return returnToken;
                             }
+
+                            // Casts that result to TypeCompareState::May and involve a ByRefLike type,
+                            // must be checked at run-time.
+                            if (opts == BoxPatterns::IsByRefLike)
+                            {
+                                assert(castResult == TypeCompareState::May);
+                                JITDUMP(
+                                    "\n Importing BOX; ISINST; BR_TRUE/FALSE for ByRefLike, must be checked at run-time\n");
+
+                                impSpillSideEffects(false, CHECK_SPILL_ALL DEBUGARG("spilling side-effects"));
+                                impPopStack();
+
+                                GenTree* toType   = impTokenToHandle(pResolvedToken);
+                                GenTree* fromType = impTokenToHandle(&isInstResolvedToken);
+                                impPushOnStack(gtNewHelperCallNode(CORINFO_HELP_ISASSIGNABLE, TYP_INT, toType,
+                                                                   fromType),
+                                               typeInfo(TYP_INT));
+                                return returnToken;
+                            }
```

Comment:
Does the behavior of this helper differ from the fallback case where we did not recognize the pattern here and just ended up in the standard `isinst` importation path?

Reply:
Yes, in the sense that it is narrower and doesn't result in a deoptimization in `impCastClassOrIsInstToTree`. Instead the result is practically identical to the optimization path, but even narrower because it is exclusively for the `BoxPatterns::IsByRefLike` case.

Reply:
Gotcha, I think I understand now.